### PR TITLE
cross-document error propagation

### PR DESCRIPTION
**Pull request by turbocrime**
should now convey serialized errors all the way to the client in every situation

no size diff

Screenshot 2024-03-13 at 17 17 59

should help with #751 but there will be another underlying cause

### CODE REVIEW

**Review by jessepinho — CHANGES_REQUESTED** [review]
Hey, this seems to introduce regressions from https://github.com/penumbra-zone/web/pull/742 (which was merged earlier today). Can you please make sure those 3 cases are working first?

**Comment by turbocrime** [comment]
okay i got confused by the broken HMR :[ i thought i wasn't seeing the issue, because it kinda works depending on where you edit things

wip pushed up i'll work on it tomorrow. annoyingly i think errors are propagating but they're getting detyped when they cross the dom message channel

File: apps/extension/src/message/popup.ts
Comment (turbocrime) [inline apps/extension/src/message/popup.ts:1]: popup responses may be absent

File: apps/extension/src/entry/offscreen-handler.ts
Comment (turbocrime) [inline apps/extension/src/entry/offscreen-handler.ts:1]: offscreen now uses ConnectError to serialize failures, and handles more failures

File: apps/extension/src/popup.ts
Comment (turbocrime) [inline apps/extension/src/popup.ts:1]: popup utility catches a closed window into an absent response

File: apps/extension/src/state/tx-approval.ts
Comment (turbocrime) [inline apps/extension/src/state/tx-approval.ts:1]: slice logic no longer necessary

File: apps/minifront/src/state/helpers.ts
Comment (turbocrime) [inline apps/minifront/src/state/helpers.ts:1]: we don't have `ConnectError` in this scope, so still doing string detection. i think if we want to detect actual codes instead of the code annotated in the message, that should be a separate PR

File: packages/router/src/grpc/offscreen-client.ts
Comment (turbocrime) [inline packages/router/src/grpc/offscreen-client.ts:1]: maybe unnecessary to change, but racing every build result with cancel was excessive and caused more throws than necessary

File: packages/transport-chrome/session-client.ts
Comment (turbocrime) [inline packages/transport-chrome/session-client.ts:1]: now using ConnectError in content scripts

File: packages/transport-dom/src/create.ts [hunk 1]
```diff
@@ -164,17 +174,15 @@ export const createChannelTransport = ({
...
-        const [{ value }, { done }] = [
-          (await iter.next()) as IteratorYieldResult<PartialMessage<I>>,
-          await iter.next(),
-        ];
-        if (!done)
+        const [{ value } = { value: null }, { done }] = [await iter.next(), await iter.next()];
+        if (done && typeof value === 'object' && value != null) {
+          const message = Any.pack(new method.I(value as object)).toJson(jsonOptions);
+          port.postMessage({ requestId, message, header } satisfies TransportMessage);
+        } else
```
Comment (turbocrime) [inline packages/transport-dom/src/create.ts:181]: this removes the need for a very specific cast by using a default value, and then using a safer cast. the expected behavior is now confirmed with tests

File: apps/extension/src/popup.ts [hunk 2]
```diff
@@ -45,7 +53,9 @@ const spawnDetachedPopup = async (path: string) => {
 const throwIfAlreadyOpen = (path: string) =>
   chrome.runtime
     .getContexts({
-      documentUrls: [chrome.runtime.getURL(path)],
+      documentUrls: [
+        path.startsWith(chrome.runtime.getURL('')) ? path : chrome.runtime.getURL(path),
```
Comment (jessepinho) [inline apps/extension/src/popup.ts:57]: What's this do? / why this change?

File: packages/transport-chrome/session-client.ts [hunk 3]
```diff
@@ -60,13 +61,10 @@ export class CRSessionClient {
...
-    this.clientPort.addEventListener('message', (ev: MessageEvent<unknown>) => {
-      if (isTransportEvent(ev.data)) {
-        const { requestId } = ev.data;
-        this.clientPort.postMessage({ requestId, error: 'Connection closed' });
-      }
+    this.clientPort.postMessage({
+      error: errorToJson(new ConnectError('Connection closed', Code.Unavailable), undefined),
     });
-    this.clientPort.postMessage({ error: { code: 'unavailable', message: 'Connection closed' } });
+    this.clientPort.close();
```
Comment (jessepinho) [inline packages/transport-chrome/session-client.ts:67]: Why this change?

File: apps/extension/src/popup.ts [hunk 4]
```diff
@@ -6,14 +6,22 @@ import type {
...
-  return chrome.runtime.sendMessage<InternalRequest<M>, InternalResponse<M>>(req);
+  await new Promise(resolve => setTimeout(resolve, 800));
+  return chrome.runtime.sendMessage<InternalRequest<M>, InternalResponse<M>>(req).catch(e => {
+    if (isChromeResponderDroppedError(e))
+      return {
+        type: req.type,
+        data: null,
+      };
+    else throw e;
+  });
```
Comment (jessepinho) [inline apps/extension/src/popup.ts:24]: I get what's happening here, but not how it fits into the PR as a whole

**Review by jessepinho — COMMENTED** [review]
Hey so, again, there's some very specific niche knowledge in here about messaging that makes it impossible to really review this without a more thorough PR description or inline comments. I added a few questions inline — if you can explain those, then I can give it a final review. But _please_ do this before requesting a review in the future, because otherwise I have to do multiple rounds of reviews.

File: apps/extension/src/popup.ts (hunk 2 above)
Comment (turbocrime) [reply apps/extension/src/popup.ts:57]: left a pr comment here that may have been concealed by gh - the previous logic was failing to match paths because it would unnecessarily append an extension origin to a path that already contained it

File: packages/transport-chrome/session-client.ts (hunk 3 above)
Comment (turbocrime) [reply packages/transport-chrome/session-client.ts:67]: using connecterror now in this script. the lines immediately above this are conveying a transport-level error that is equivalent to this

https://github.com/penumbra-zone/web/blob/0cc03146874e28e8fff14046e5ba04a084f1f533/packages/transport-chrome/session-client.ts#L62-L68

previously this also attached a handler that closed every subsequent request attempt. but, any existing transports are failed by the close message, and no subsequent transports can init.

File: apps/extension/src/popup.ts (hunk 4 above)
Comment (turbocrime) [reply apps/extension/src/popup.ts:24]: this pr is addressing cross-document error propagation.

the closing of the popup document surfaces as an error, but it is not a failure, it is a meaningful response. handling the termination of a document from within the document is complex, and we shouldn't reimplement it for every kind of popup, so it's handled generally for all popups.

**Review by jessepinho — APPROVED** [review]
ship it!
